File: src/release.py
import re
import json
from sys import exit
from pathlib import Path
from src import repository, gh

def convert_title(text):
    if not text or not isinstance(text, str):
        return text
    return re.sub(
        r'\b([a-z0-9]+(?:-[a-z0-9]+)*)\b',
        lambda m: m.group(1).replace('-', ' ').title(),
        text,
        flags=re.IGNORECASE
    )

def extract_version(file_path):
    if not file_path:
        return 'unknown'
    path = Path(file_path)
    base_name = path.stem
    match = re.search(r'(\d+\.\d+\.\d+(-[a-z]+\.\d+)?(-release\d*)?)', base_name)
    return match.group(1) if match else 'unknown'

def create_github_release(name, patches_name, cli_name, apk_file_path):
    patchver = extract_version(patches_name)
    cliver = extract_version(cli_name)
    tag_name = f"{name}-v{patchver}"

    apk_path = Path(apk_file_path)
    if not apk_path.exists():
        exit(1)

    repo = gh.get_repo(repository)

    # Step 1: Check for existing release with the exact tag name
    try:
        existing_release = repo.get_release(tag_name)
    except:
        existing_release = None

    # Step 2: Delete existing assets if same APK already uploaded
    if existing_release:
        for asset in existing_release.get_assets():
            if asset.name == apk_path.name:
                asset.delete_asset()

    # Step 3: Delete old releases with the same base name and matching version suffix
    releases = list(repo.get_releases())

    suffix_match = re.search(r'(-[a-z]+\.\d+)$', patchver)
    current_suffix = suffix_match.group(1) if suffix_match else ''

    for release in releases:
        release_tag = release.tag_name
        if release_tag.startswith(f"{name}-v") and release_tag != tag_name:
            old_version = release_tag[len(name) + 2:]
            old_suffix_match = re.search(r'(-[a-z]+\.\d+)$', old_version)
            old_suffix = old_suffix_match.group(1) if old_suffix_match else ''

            if old_suffix == current_suffix:
                old_numeric = re.sub(r'(-[a-z]+\.\d+)?(-release\d*)?$', '', old_version)
                current_numeric = re.sub(r'(-[a-z]+\.\d+)?(-release\d*)?$', '', patchver)
                if old_numeric < current_numeric:
                    release.delete_release()

    # Step 4: Create new release if it doesn't exist
    if not existing_release:
        release_body = f"""\
# Release Notes

## Build Tools:
- **ReVanced Patches:** v{patchver}
- **ReVanced CLI:** v{cliver}

## Note:
**ReVanced GmsCore** is **necessary** to work. 
- Please **download** it from [HERE](https://github.com/revanced/gmscore/releases/latest).
"""
        release_name = f"{convert_title(name)} v{patchver}"
        existing_release = repo.create_git_release(
            tag=tag_name,
            name=release_name,
            message=release_body,
            draft=False,
            prerelease=False
        )

    # Step 5: Upload APK
    existing_release.upload_asset(
        path=str(apk_path),
        label=apk_path.name,
        content_type='application/vnd.android.package-archive'
    )

File: lib/release.js
const fs = require("fs");
const path = require("path");
const https = require("https");

const { downloadLatestGithubAsset } = require("./github");

const TOKEN = process.env.GITHUB_TOKEN;
const REPO = process.env.GITHUB_REPOSITORY;

function request(options, body) {
  return new Promise((resolve, reject) => {
    const req = https.request(options, (res) => {
      let data = "";
      res.on("data", (d) => (data += d));
      res.on("end", () => {
        try {
          resolve(JSON.parse(data));
        } catch {
          resolve(data);
        }
      });
    });
    req.on("error", reject);
    if (body) req.write(body);
    req.end();
  });
}

const headers = {
  "User-Agent": "node",
  Authorization: `Bearer ${TOKEN}`,
  Accept: "application/vnd.github+json",
};

// ===== Release =====
async function getOrCreateRelease(tag) {
  let res = await request({
    hostname: "api.github.com",
    path: `/repos/${REPO}/releases/tags/${tag}`,
    method: "GET",
    headers,
  });

  if (res.message === "Not Found") {
    console.log("🆕 Creating release:", tag);

    res = await request(
      {
        hostname: "api.github.com",
        path: `/repos/${REPO}/releases`,
        method: "POST",
        headers,
      },
      JSON.stringify({
        tag_name: tag,
        name: tag,
        draft: false,
        prerelease: false,
      })
    );
  }

  return res;
}

async function getAssets(releaseId) {
  return request({
    hostname: "api.github.com",
    path: `/repos/${REPO}/releases/${releaseId}/assets`,
    method: "GET",
    headers,
  });
}

async function deleteAsset(id) {
  return request({
    hostname: "api.github.com",
    path: `/repos/${REPO}/releases/assets/${id}`,
    method: "DELETE",
    headers,
  });
}

async function upload(uploadUrl, filePath) {
  const fileName = path.basename(filePath);
  const data = fs.readFileSync(filePath);

  const url = new URL(
    uploadUrl.replace("{?name,label}", `?name=${encodeURIComponent(fileName)}`)
  );

  return request(
    {
      hostname: url.hostname,
      path: url.pathname + url.search,
      method: "POST",
      headers: {
        ...headers,
        "Content-Type": "application/vnd.android.package-archive",
        "Content-Length": data.length,
      },
    },
    data
  );
}

// ===== Upload helper (replace if exists) =====
async function uploadWithReplace(release, filePath) {
  const fileName = path.basename(filePath);

  const assets = await getAssets(release.id);
  const exist = assets.find(a => a.name === fileName);

  if (exist) {
    console.log("♻️ Replace:", fileName);
    await deleteAsset(exist.id);
  }

  console.log("📤 Upload:", fileName);
  return upload(release.upload_url, filePath);
}

// ===== MAIN =====
async function uploadApkRelease({ version, apkPath }) {
  if (!TOKEN) throw new Error("Missing GITHUB_TOKEN");
  if (!REPO) throw new Error("Missing GITHUB_REPOSITORY");

  const tag = `youtube-${version}`;
  const release = await getOrCreateRelease(tag);

  // 1. Upload patched APK
  await uploadWithReplace(release, apkPath);

  // 2. Download + upload MicroG (reuse github.js)
  console.log("📦 Fetch MicroG...");
  const microgPath = await downloadLatestGithubAsset({
    owner: "MorpheApp",
    repo: "MicroG-RE",
    match: (n) => n.endsWith(".apk"),
  });

  await uploadWithReplace(release, microgPath);

  console.log("✅ Release done");
}

module.exports = { uploadApkRelease };
